addons/sale/migrations/9.0.1.0/pre-migration.py

```diff
@@ -56,9 +56,83 @@ def map_order_state(cr):
         WHERE sol.order_id = so.id""")
 
 
+def prepopulate_fields(cr):
+    """ Recomputing a fields will be very expensive via the ORM, so do it
+    here for fields where the computation is trivial"""
+    cr.execute('alter table sale_order_line add column qty_invoiced numeric')
+    cr.execute(
+        """\
+        with line2qty_invoiced(id, qty_invoiced) as (
+            select l.id, sum(
+                case
+                    when i.type='out_invoice' then il.quantity
+                    when i.type='out_refund' then -il.quantity
+                end * uom_invoice.factor / uom_sale.factor
+            ) s
+            from sale_order_line l
+            join sale_order_line_invoice_rel r on r.order_line_id=l.id
+            join account_invoice_line il on r.invoice_line_id=il.id
+            join account_invoice i on i.id=il.invoice_id
+            join product_uom uom_invoice on il.uom_id=uom_invoice.id
+            join product_uom uom_sale on l.product_uom=uom_sale.id
+            group by l.id
+        )
+        update sale_order_line
+        set qty_invoiced=line2qty_invoiced.qty_invoiced
+        from line2qty_invoiced where line2qty_invoiced.id=sale_order_line.id
+        """)
+    cr.execute('alter table sale_order_line add column qty_to_invoice numeric')
+    cr.execute(
+        """\
+        update sale_order_line
+        set qty_to_invoice=product_uom_qty - qty_invoiced
```

Comment:
@hbrunn This sql sentence is not correct. It does not replicate the same conditions as https://github.com/odoo/odoo/blob/9.0/addons/sale/sale.py#L516

For example, in products where the invoice policy is set based on delivered quantities the quantity to invoice is to be based on the quantity delivered - quantity invoiced.